jsettlers.logic/src/jsettlers/ai/highlevel/AiStatistics.java

```diff
@@ -0,0 +1,612 @@
+import static jsettlers.common.landscape.ELandscapeType.RIVER1;
+import static jsettlers.common.landscape.ELandscapeType.RIVER2;
+import static jsettlers.common.landscape.ELandscapeType.RIVER3;
+import static jsettlers.common.landscape.ELandscapeType.RIVER4;
+import static jsettlers.common.mapobject.EMapObjectType.STONE;
+import static jsettlers.common.mapobject.EMapObjectType.TREE_ADULT;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L1;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L2;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L3;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+import jsettlers.algorithms.construction.AbstractConstructionMarkableMap;
+import jsettlers.common.CommonConstants;
+import jsettlers.common.buildings.EBuildingType;
+import jsettlers.common.buildings.IBuilding;
+import jsettlers.common.landscape.ELandscapeType;
+import jsettlers.common.landscape.EResourceType;
+import jsettlers.common.map.shapes.MapCircle;
+import jsettlers.common.mapobject.EMapObjectType;
+import jsettlers.common.material.EMaterialType;
+import jsettlers.common.movable.EMovableType;
+import jsettlers.common.position.ShortPoint2D;
+import jsettlers.logic.buildings.Building;
+import jsettlers.logic.map.grid.MainGrid;
+import jsettlers.logic.map.grid.flags.FlagsGrid;
+import jsettlers.logic.map.grid.landscape.LandscapeGrid;
+import jsettlers.logic.map.grid.movable.MovableGrid;
+import jsettlers.logic.map.grid.objects.ObjectsGrid;
+import jsettlers.logic.map.grid.partition.PartitionsGrid;
+import jsettlers.logic.movable.Movable;
+import jsettlers.logic.objects.stack.StackMapObject;
+import jsettlers.logic.player.Player;
+
+/**
+ * This class calculates statistics based on the grids which are used by highlevel and lowlevel KI. The statistics are calculated once and read
+ * multiple times within one AiExecutor step triggerd by the game clock.
+ * 
+ * @author codingberlin
+ */
+public class AiStatistics {
+
+	private static final short BORDER_LAND_WIDTH = 10;
+
+	private final Queue<Building> buildings;
+	private Map<Integer, Map<EBuildingType, Integer>> totalBuildingsNumbers;
+	private Map<Integer, Map<EBuildingType, Integer>> buildingsNumbers;
+	private Map<Integer, Map<EBuildingType, Integer>> unoccupiedBuildingsNumbers;
+	private Map<Integer, Map<EMaterialType, Integer>> materialNumbers;
+	private Map<Integer, Map<EMovableType, List<ShortPoint2D>>> movablePositions;
+	private Map<Integer, Integer> numberOfNotFinishedBuildings;
+	private Map<Integer, Integer> numberOfTotalBuildings;
+	private Map<Integer, Integer> numberOfNotOccupiedTowers;
+	private Map<Integer, List<ShortPoint2D>> stones;
+	private Map<Integer, List<ShortPoint2D>> trees;
+	private Map<Integer, List<ShortPoint2D>> rivers;
+	private Map<EMapObjectType, Map<Integer, List<Integer>>> sortedCuttableObjectsInDefaultPartition;
```

Comment:
this sorted* lists are used to do a "nearest neighbor search" algorithm which does not visit every member in a list to reach a lower complexity as O(n*n)